lib/index.js

```diff
@@ -3055,6 +3059,254 @@ UFDS.prototype.listResolvers = function listResolvers(region, cb, noCache) {
+ * @throws {TypeError} on bad input.
+ */
+UFDS.prototype.addAccessKey = function addAccessKey(user, account, cb) {
+    if (typeof (account) === 'function') {
+        cb = account;
+        account = null;
+    }
+    if (typeof (user) !== 'string') {
+        assert.object(user, 'user');
+        if (user.account) {
+            account = user.account;
+        }
+    }
+    assert.func(cb, 'callback');
+    account = workAroundUnderspecification(account, 'account');
+
+    cb = once(cb);
+
+    var self = this;
+
+    function _addAccKey(init_err, usr) {
+        if (init_err) {
+            cb(init_err);
+            return;
+        }
+
+        var entry = {
+            objectclass: 'accesskey',
+            created: Date.now()
+        };
+
+        try {
+            var accessKeyId = crypto.randomBytes(16).toString('hex');
+            var accessKeySecret = crypto.randomBytes(32).toString('hex');
+            entry.accesskeyid = accessKeyId;
+            entry.accesskeysecret = accessKeySecret;
+        } catch (e) {
+            cb(new InternalError(e.message));
+            return;
+        }
+
+        var dn = (account) ?
+            sprintf(SUBUSER_ACCESS_KEY_FMT,
+                entry.accesskeyid, usr.uuid, account) :
+            sprintf(USER_ACCESS_KEY_FMT, entry.accesskeyid, usr.uuid);
+
+        self.add(dn, entry, function (err) {
+            if (err) {
+                cb(translateError(err));
+                return;
+            }
+            self.getAccessKey(usr, entry.accesskeyid, account, cb);
+        });
+    }
+
+    if (typeof (user) === 'object') {
+        _addAccKey(null, user);
+    } else {
+        this.getUser(user, account, _addAccKey);
+    }
+};
+
+
+/**
+ * Retrieves an AccessKey by Id.
+ *
+ * @param {Object} user the object you got back from getUser.
+ * @param {String} AccessKeyId of the AccessKey you want.
+ * @param {String} (Optional) account uuid for a customer sub-user
+ * @param {Function} callback of the form fn(err, key).
+ * @throws {TypeError} on bad input.
+ */
+UFDS.prototype.getAccessKey =
+function getAccessKey(user, id, account, cb, noCache) {
+    if (typeof (account) === 'function') {
+        cb = account;
+        account = null;
+    }
+    if (typeof (user) !== 'string') {
+        assert.object(user, 'user');
+        if (user.account) {
+            account = user.account;
+        }
+    }
+    assert.string(id, 'accesskeyid');
+    assert.func(cb, 'callback');
+    account = workAroundUnderspecification(account, 'account');
+
+    cb = once(cb);
+
+    var self = this;
+    function _accKeySearch(err, usr) {
+        if (err) {
+            cb(err);
+            return;
+        }
+        var filter = util.format(
+            '(&(objectclass=accesskey)(accesskeyid=%s))', id);
+        var opts = {
+            scope: 'one',
+            filter: filter
+        };
+        self.search(usr.dn, opts, function (sErr, keys) {
```

Comment:
Maybe assert `usr.dn`

Reply:
Yeah, it'll be present for sure in the case we pass an UUID, but not when an object is given. Added.